# Tarefa 02 Módulo 05

O nosso projeto desta sequência de módulos do curso será um aprofundamento da demonstração sobre classificação de risco de crédito que vimos lá no comecinho. Pois recebemos uma base já montada pra nós. Tenha certeza de que ela passou por um longo processamento até ficar daquele jeito. Neste exercício vamos exercitar o que aprendemos nas ultimas aulas e montar a variável resposta da base do nosso projeto.

#### Marcação de bom e mau
O objetivo da modelagem é classificar o risco de inadimplência, ou como se diz no meio, o risco de *default*. Podemos fazer longas discussões sobre o conceito de *default* com base em estudos e exigências regulatórias, para efeitos deste estudo, um cliente em *default* é aquele que está em 60 dias de atraso ou mais. Então classificaremos os clientes como 'bons' e 'maus' assim:
- **Maus** pagadores: são aqueles que entraram em 'default' (atraso 60 dias ou mais) nos 24 meses seguintes à aquisição do cartão de crédito.
- **Bons** pagadores: são considerados todos os demais.
- **Excluídos**: Clientes que não adquiriram um cartão de crédito (seja por recusa, seja por desistência) não possuem informações de pagamento, portanto não se pode identificar se são bons ou maus. Há uma longa discussão e literatura sobre *inferência de rejeitados* que está fora do escopo deste exercício.

#### Bases disponíveis
Temos duas bases importantes aqui: uma de propostas, com diversas informações dos vários solicitantes de cartão de crédito, e uma base de pagamentos. A base de pagamentos será utilizada para identificar a ocorrência de *default*. A base de propostas tem diversas informações coletadas no momento da solicitação do crédito (isto é importante: qualquer informação posterior a essa data é impossível de ser coletada na aplicação do modelo e não pode ser utilizada).

As variáveis delas são:

Base de propostas - application_records.csv

| Nome da Variável         | Description                                         | Tipo  |
| ------------------------ |:---------------------------------------------------:| -----:|
| ID| identificador do cliente (chave) |inteiro|
| CODE_GENDER| M = 'Masculino'; F = 'Feminino' |M/F|
| FLAG_OWN_CAR| Y = 'possui'; N = 'não possui' |Y/N|
| FLAG_OWN_REALTY| Y = 'possui'; N = 'não possui' |Y/N|
| CNT_CHILDREN| Quantidade de filhos |inteiro|
| AMT_INCOME_TOTAL| Annual income |inteiro|
| NAME_INCOME_TYPE|Tipo de renda (ex: assaliariado, autônomo etc) | texto |
| NAME_EDUCATION_TYPE| Nível de educação (ex: secundário, superior etc) |texto|
| NAME_FAMILY_STATUS | Estado civil (ex: solteiro, casado etc)| texto |
| NAME_HOUSING_TYPE | tipo de residência (ex: casa/apartamento, com os pais etc) | texto |
| DAYS_BIRTH | Count backwards from current day (0), -1 means yesterday |inteiro|
| DAYS_EMPLOYED | Count backwards from current day (0), -1 means yesterday |inteiro|
| FLAG_MOBIL | Indica se possui celular (1 = sim, 0 = não) |binária|
| FLAG_WORK_PHONE | Indica se possui telefone comercial (1 = sim, 0 = não) |binária|
| FLAG_PHONE | Indica se possui telefone (1 = sim, 0 = não) |binária|
| FLAG_EMAIL | Indica se possui e-mail (1 = sim, 0 = não) |binária|
| OCCUPATION_TYPE | Occupation	 |Qualitativa|
| CNT_FAM_MEMBERS | quantidade de pessoas na residência |inteiro|

Base de pagamentos - pagamentos_largo.csv  

| Nome da Variável         | Description                                         | Tipo  |
| ------------------------ |:---------------------------------------------------:| -----:|
| ID| identificador do cliente (chave) |inteiro|
| mes_00 a mes_24| faixa de atraso mês a mês do cliente <br>0: 1-29 days past due &nbsp;&nbsp;&nbsp;&nbsp; 1: 30-59 days past due <br />2: 60-89 days overdue &nbsp;&nbsp;&nbsp;&nbsp; 3: 90-119 days overdue <br /> 4: 120-149 days overdue &nbsp;&nbsp;&nbsp;&nbsp; 5: more than 150 days <br />C: paid off that month &nbsp;&nbsp;&nbsp;&nbsp; X: No loan for the month |Qualitativa|

#### Construindo a variável resposta
A base de pagamentos está em um formato de 'base larga'. Essa base possui informações de pagamentos do cliente mês a mês a partir do mês de aquisição do crédito (mês 0) até o vigésimo quarto mês após a aquisição do crédito (mês 24). Utilizaremos essa base para determinar se um proponente é considerado 'bom pagador' ou caso apresente atraso representativo, será considerado 'mau pagador'.

#### Base larga vs base longa
A base ser larga significa que há uma linha para cada cliente, e que as informações estarão nas colunas, em contraste com a 'base longa', em que haveria uma linha para cada combinação cliente/mês, uma coluna indicando o cliente, outra indicando o mês, e apenas uma coluna com a informação do atraso.

#### Tarefa 1) Marcar *default* no mês
Faça uma indicadora de se o cliente está em *default* em cada uma das marcações (mes_00 a mes_24). Dica: você pode utilizar o método ```.isin()``` do Pandas. Consulte a [documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isin.html) caso necessário.

#### Tarefa 2) 'bons' e 'maus' ao longo de todos os 24 meses de desempenho
Marque para cada cliente se ele teve pelo menos um episódio de *default* entre o mês 0 e o mês 24. Dica: o método ```sum()``` pode ajudar. Caso precise, consulte a [documentação](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sum.html) e procure pelo argumento ```axis```, você viu outros métodos que possuem esse argumento também. Tendo o número de meses em default de cada cliente, basta marcar ```True``` para todos aqueles que possuem pelo menos 1 mês em *default* e ```False``` para os demais.

#### Tarefa 3) Marcando proponentes expostos ao risco de crédito
Marcando proponentes que se tornaram tomadores: lembre-se de que clientes que não adquiriram o cartão devem ser desconsiderados. A base de pagamentos possui apenas clientes que adquiriram cartão de crédito, então você pode selecionar somente os clientes da base de propostas que se encontram na base de pagamentos.

#### Tarefa 4) Consolidando as informações
Faça uma junção das informações da base de propostas com a variável de *default* que você acabou de construir. Talvez você consiga realizar a tarefa 3 e tarefa 4 em uma única linha de código ;)

#### Tarefa 5) Verificando
Faça uma contagem dos valores do *default* que você construiu.

In [16]:
import pandas as pd

In [17]:
propostas = pd.read_csv('application_record.csv')
pagamentos = pd.read_csv('pagamentos_largo.csv')

In [18]:
pagamentos.head()

ID mes_0 mes_1 mes_10 mes_11 mes_12 mes_13 mes_14 mes_15 mes_16  ...  \
0  5001718     0     0      0      0      0      0      0      0    NaN  ...   
1  5001719     0     0      C      C      C      C      C      C      C  ...   
2  5001720     0     0      0      0      0      0      0      0      0  ...   
3  5001723     0     0    NaN    NaN    NaN    NaN    NaN    NaN    NaN  ...   
4  5001726     0     0      C      C      C      C      C      C      C  ...   

  mes_22 mes_23 mes_24 mes_3 mes_4 mes_5 mes_6 mes_7 mes_8 mes_9  
0    NaN      0    NaN     0     0     0     0   NaN     0   NaN  
1      C      C      C     C     C     C     C     C     C     C  
2      1      0      0     0     0     0     0     0     0     0  
3    NaN    NaN    NaN     0     0     0     0     0   NaN   NaN  
4      C      C      C     0     0     0     C     C     C     C  

[5 rows x 26 columns]



*   NaN -> missings


In [19]:
propostas.head()

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008804           M            Y               Y             0   
1  5008805           M            Y               Y             0   
2  5008806           M            Y               Y             0   
3  5008808           F            N               Y             0   
4  5008809           F            N               Y             0   

   AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0          427500.0               Working               Higher education   
1          427500.0               Working               Higher education   
2          112500.0               Working  Secondary / secondary special   
3          270000.0  Commercial associate  Secondary / secondary special   
4          270000.0  Commercial associate  Secondary / secondary special   

     NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0        Civil marriage   Rented apartment      -12005          -4542   
1        Civil marriage   Rented apartment      -12005          -4542   
2               Married  House / apartment      -21474          -1134   
3  Single / not married  House / apartment      -19110          -3051   
4  Single / not married  House / apartment      -19110          -3051   

   FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0           1                1           0           0             NaN   
1           1                1           0           0             NaN   
2           1                0           0           0  Security staff   
3           1                0           1           1     Sales staff   
4           1                0           1           1     Sales staff   

   CNT_FAM_MEMBERS  
0              2.0  
1              2.0  
2              2.0  
3              1.0  
4              1.0

In [20]:
pagamentos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20937 entries, 0 to 20936
Data columns (total 26 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   ID      20937 non-null  int64 
 1   mes_0   20937 non-null  object
 2   mes_1   19216 non-null  object
 3   mes_10  17455 non-null  object
 4   mes_11  16972 non-null  object
 5   mes_12  16943 non-null  object
 6   mes_13  16540 non-null  object
 7   mes_14  16525 non-null  object
 8   mes_15  16155 non-null  object
 9   mes_16  16093 non-null  object
 10  mes_17  15911 non-null  object
 11  mes_18  15630 non-null  object
 12  mes_19  15336 non-null  object
 13  mes_2   19348 non-null  object
 14  mes_20  15184 non-null  object
 15  mes_21  15076 non-null  object
 16  mes_22  14797 non-null  object
 17  mes_23  14598 non-null  object
 18  mes_24  14433 non-null  object
 19  mes_3   18925 non-null  object
 20  mes_4   18802 non-null  object
 21  mes_5   18512 non-null  object
 22  mes_6   18214 non-null



*   Os meses não estão organizados em ordem cronológica.
*   Observou-se um aumento no número de valores ausentes ('missings') ao longo das colunas.
*   O campo 'ID' está atualmente categorizado como 'int64'. No entanto, considerando que este campo é um identificador único e não um valor numérico destinado a operações aritméticas, seria melhor categorizado como 'object'.
*   Os demais campos estão categorizados como 'object'.

Linhas: 20937 Colunas: 26



In [21]:
propostas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 438557 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   ID                   438557 non-null  int64  
 1   CODE_GENDER          438557 non-null  object 
 2   FLAG_OWN_CAR         438557 non-null  object 
 3   FLAG_OWN_REALTY      438557 non-null  object 
 4   CNT_CHILDREN         438557 non-null  int64  
 5   AMT_INCOME_TOTAL     438557 non-null  float64
 6   NAME_INCOME_TYPE     438557 non-null  object 
 7   NAME_EDUCATION_TYPE  438557 non-null  object 
 8   NAME_FAMILY_STATUS   438557 non-null  object 
 9   NAME_HOUSING_TYPE    438557 non-null  object 
 10  DAYS_BIRTH           438557 non-null  int64  
 11  DAYS_EMPLOYED        438557 non-null  int64  
 12  FLAG_MOBIL           438557 non-null  int64  
 13  FLAG_WORK_PHONE      438557 non-null  int64  
 14  FLAG_PHONE           438557 non-null  int64  
 15  FLAG_EMAIL       

*    A variável 'ID', está categorizada como 'int64'. No entanto, a natureza deste campo sugere que seria mais adequado tipá-lo como 'object'.

Linhas: 211538  Colunas: 18

In [23]:
pagamentos.isnull().sum()

ID           0
mes_0        0
mes_1     1721
mes_10    3482
mes_11    3965
mes_12    3994
mes_13    4397
mes_14    4412
mes_15    4782
mes_16    4844
mes_17    5026
mes_18    5307
mes_19    5601
mes_2     1589
mes_20    5753
mes_21    5861
mes_22    6140
mes_23    6339
mes_24    6504
mes_3     2012
mes_4     2135
mes_5     2425
mes_6     2723
mes_7     3175
mes_8     3196
mes_9     3427
dtype: int64

In [27]:
propostas.isnull().sum()

ID                          0
CODE_GENDER                 0
FLAG_OWN_CAR                0
FLAG_OWN_REALTY             0
CNT_CHILDREN                0
AMT_INCOME_TOTAL            0
NAME_INCOME_TYPE            0
NAME_EDUCATION_TYPE         0
NAME_FAMILY_STATUS          0
NAME_HOUSING_TYPE           0
DAYS_BIRTH                  0
DAYS_EMPLOYED               0
FLAG_MOBIL                  0
FLAG_WORK_PHONE             0
FLAG_PHONE                  0
FLAG_EMAIL                  0
OCCUPATION_TYPE        134203
CNT_FAM_MEMBERS             0
dtype: int64

In [31]:
pagamentos_com_na = pagamentos.shape[0]
propostas_com_na = propostas.shape[0]

print(f"Pagamentos: {pagamentos_com_na}.")
print(f"Propostas: {propostas_com_na}.")

Pagamentos: 20937.
Propostas: 438557.


In [32]:
pagamentos.dropna(inplace = True)
pagamentos_sem_na = pagamentos.shape[0]

propostas.dropna(inplace = True)
propostas_sem_na = propostas.shape[0]

print(f"Propostas: {propostas_sem_na}.")
print(f"Pagamentos: {pagamentos_sem_na}.")

Propostas: 304354.
Pagamentos: 12267.


In [46]:
print(f"Linhas removidas do dataframe 'pagamentos': {pagamentos_com_na - pagamentos_sem_na}.")
print(f"Representando {round(((pagamentos_com_na - pagamentos_sem_na)/20936) * 100)}% das linhas.")
print(f"Linhas removidas do dataframe 'propostas': {propostas_com_na - propostas_sem_na}.")
print(f"Representando {round(((propostas_com_na - propostas_sem_na)/438556) * 100)}% das linhas.")

Linhas removidas do dataframe 'pagamentos': 8670.
Representando 41% das linhas.
Linhas removidas do dataframe 'propostas': 134203.
Representando 31% das linhas.


In [47]:
pagamentos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12267 entries, 1 to 20936
Data columns (total 26 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   ID      12267 non-null  int64 
 1   mes_0   12267 non-null  object
 2   mes_1   12267 non-null  object
 3   mes_10  12267 non-null  object
 4   mes_11  12267 non-null  object
 5   mes_12  12267 non-null  object
 6   mes_13  12267 non-null  object
 7   mes_14  12267 non-null  object
 8   mes_15  12267 non-null  object
 9   mes_16  12267 non-null  object
 10  mes_17  12267 non-null  object
 11  mes_18  12267 non-null  object
 12  mes_19  12267 non-null  object
 13  mes_2   12267 non-null  object
 14  mes_20  12267 non-null  object
 15  mes_21  12267 non-null  object
 16  mes_22  12267 non-null  object
 17  mes_23  12267 non-null  object
 18  mes_24  12267 non-null  object
 19  mes_3   12267 non-null  object
 20  mes_4   12267 non-null  object
 21  mes_5   12267 non-null  object
 22  mes_6   12267 non-null  obj

In [59]:
# 1) Seu código aqui
#'2', '3', '4', '5'
default = pagamentos.isin(['2', '3', '4', '5'])

default_qtde = pagamentos.any(axis = 1).sum()

print(f"Quantidade de default: {default_qtde}.")

Quantidade de default: 12267.


In [62]:
# 2) Seu código aqui
pagamentos['default'] = default.any(axis=1)
clientes_default = pagamentos['default'].sum()
print(f"Quantidade de clientes 'deafult': {clientes_default}.")

Quantidade de clientes 'deafult': 299.


In [64]:
# 3) Seu código aqui
propostas_pagamentos = propostas.merge(pagamentos, on = 'ID', how = 'inner')
proponentes_risco_credito = propostas_pagamentos.shape[0]
print(f"Quantidade de proponentes que se tornaram tomadores: {proponentes_risco_credito}.")

Quantidade de proponentes que se tornaram tomadores: 6334.


In [ ]:
# 4) eu código aqui